From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: csagan5 <32685696+csagan5@users.noreply.github.com>
Date: Fri, 24 Jun 2022 22:24:22 +0200
Subject: Revert removal of execution context address space

Revert "[Private Network Access] Remove ExecutionContext::SetAddressSpace()."
This reverts commit dfa3a4a6c0d023e5126dad31f6c479dc8d20e416.

Revert "[Private Network Access] Remove ExecutionContext::AddressSpace()."
This reverts commit a5c64b9f5f985cfebab3375b53062321a62e3e31.

Revert "[Private Network Access] Remove GlobalScopeCreationParams address space."
This reverts commit c32df617780283942d49ca5574fcd6b7f691e9c1.

Revert "[Private Network Access] Remove WorkerClassicScriptLoader address space."
This reverts commit 11f6a3e29bee9a6590a4fd56b28e8048200c18d0.

Original License: BSD-3-Clause - https://spdx.org/licenses/BSD-3-Clause.html
License: GPL-3.0-only - https://spdx.org/licenses/GPL-3.0-only.html
---
 .../core/execution_context/execution_context.cc  |  9 +++++++++
 .../core/execution_context/execution_context.h   |  4 ++++
 .../core/exported/web_shared_worker_impl.cc      |  1 +
 .../renderer/core/frame/policy_container.cc      | 10 ++++++++++
 .../blink/renderer/core/frame/policy_container.h | 11 +++++++++++
 .../custom/layout_worklet_global_scope_proxy.cc  |  1 +
 ...alled_service_worker_module_script_fetcher.cc |  1 +
 .../modulescript/module_script_loader_test.cc    |  1 +
 .../modulescript/worker_module_script_fetcher.cc |  3 ++-
 .../renderer/core/workers/dedicated_worker.cc    |  9 ++++++++-
 .../renderer/core/workers/dedicated_worker.h     |  3 +++
 .../workers/dedicated_worker_global_scope.cc     |  9 ++++++++-
 .../core/workers/dedicated_worker_global_scope.h |  1 +
 .../core/workers/global_scope_creation_params.cc |  3 +++
 .../core/workers/global_scope_creation_params.h  |  7 +++++++
 .../core/workers/installed_scripts_manager.cc    | 16 ++++++++++++++++
 .../core/workers/installed_scripts_manager.h     |  5 +++++
 .../core/workers/main_thread_worklet_test.cc     |  1 +
 .../core/workers/shared_worker_global_scope.cc   |  7 +++++++
 .../core/workers/shared_worker_global_scope.h    |  1 +
 .../workers/threaded_worklet_messaging_proxy.cc  |  5 ++++-
 .../core/workers/threaded_worklet_test.cc        |  1 +
 .../core/workers/worker_classic_script_loader.cc |  6 +++++-
 .../core/workers/worker_classic_script_loader.h  |  6 ++++++
 .../renderer/core/workers/worker_global_scope.h  |  2 ++
 .../renderer/core/workers/worker_thread_test.cc  |  1 +
 .../core/workers/worker_thread_test_helper.h     |  5 +++++
 .../csspaint/paint_worklet_global_scope_proxy.cc |  2 +-
 .../modules/exported/web_embedded_worker_impl.cc |  1 +
 .../service_worker_global_scope.cc               | 13 +++++++++++--
 .../service_worker/service_worker_global_scope.h |  2 ++
 .../webaudio/audio_worklet_global_scope_test.cc  |  1 +
 .../webaudio/audio_worklet_thread_test.cc        |  1 +
 .../worklet/worklet_thread_test_common.cc        |  2 +-
 34 files changed, 142 insertions(+), 9 deletions(-)

diff --git a/third_party/blink/renderer/core/execution_context/execution_context.cc b/third_party/blink/renderer/core/execution_context/execution_context.cc
index 7c1735534e10445cca3ad4eb42c0c02f3c643eef..f87746aebc816bc14b74623145888799f6b12bcd 100644
--- a/third_party/blink/renderer/core/execution_context/execution_context.cc
+++ b/third_party/blink/renderer/core/execution_context/execution_context.cc
@@ -524,6 +524,15 @@ void ExecutionContext::SetReferrerPolicy(
   SetInitiatorStateToken(new_initiator_state_token);
 }
 
+network::mojom::IPAddressSpace ExecutionContext::AddressSpace() const {
+  return policy_container_->GetIPAddressSpace();
+}
+
+void ExecutionContext::SetAddressSpace(
+    network::mojom::blink::IPAddressSpace ip_address_space) {
+  GetPolicyContainer()->SetIPAddressSpace(ip_address_space);
+}
+
 void ExecutionContext::SetPolicyContainer(
     std::unique_ptr<PolicyContainer> container) {
   policy_container_ = std::move(container);
diff --git a/third_party/blink/renderer/core/execution_context/execution_context.h b/third_party/blink/renderer/core/execution_context/execution_context.h
index 4ca0f55dcec5736ecef0a2d43043b5ea54820274..d5b78149cc1b444aff4cd260919edfba44f5a66e 100644
--- a/third_party/blink/renderer/core/execution_context/execution_context.h
+++ b/third_party/blink/renderer/core/execution_context/execution_context.h
@@ -34,6 +34,7 @@
 #include "net/storage_access_api/status.h"
 #include "services/metrics/public/cpp/ukm_source_id.h"
 #include "services/network/public/mojom/permissions_policy/permissions_policy_feature.mojom-blink-forward.h"
+#include "services/network/public/mojom/ip_address_space.mojom-blink-forward.h"
 #include "services/network/public/mojom/referrer_policy.mojom-blink-forward.h"
 #include "third_party/blink/public/common/fingerprinting_protection/noise_token.h"
 #include "third_party/blink/public/common/tokens/tokens.h"
@@ -397,6 +398,9 @@ class CORE_EXPORT ExecutionContext : public Supplementable<ExecutionContext>,
       const String& message = g_empty_string,
       const String& source_file = g_empty_string) const {}
 
+  network::mojom::IPAddressSpace AddressSpace() const;
+  void SetAddressSpace(network::mojom::blink::IPAddressSpace ip_address_space);
+
   HeapObserverList<ContextLifecycleObserver>& ContextLifecycleObserverSet();
   unsigned ContextLifecycleStateObserverCountForTesting() const;
 
diff --git a/third_party/blink/renderer/core/exported/web_shared_worker_impl.cc b/third_party/blink/renderer/core/exported/web_shared_worker_impl.cc
index a3a718f49e9663d22ab25ee157dc7f3f1e33d4f0..70ebc445800a2f91e08f0485863f183b6aa640f1 100644
--- a/third_party/blink/renderer/core/exported/web_shared_worker_impl.cc
+++ b/third_party/blink/renderer/core/exported/web_shared_worker_impl.cc
@@ -285,6 +285,7 @@ void WebSharedWorkerImpl::StartWorkerContext(
       MakeGarbageCollected<WorkerClients>(),
       std::make_unique<SharedWorkerContentSettingsProxy>(
           std::move(content_settings)),
+      std::nullopt /* response_address_space */,
       nullptr /* inherited_trial_features */, devtools_worker_token,
       std::move(worker_settings), mojom::blink::V8CacheOptions::kDefault,
       nullptr /* worklet_module_response_map */,
diff --git a/third_party/blink/renderer/core/frame/policy_container.cc b/third_party/blink/renderer/core/frame/policy_container.cc
index 31b8d3e179e334129d767b2422f3940faadd8597..f8fda8065b1dab515b63d6a8073935bbbb3d4c89 100644
--- a/third_party/blink/renderer/core/frame/policy_container.cc
+++ b/third_party/blink/renderer/core/frame/policy_container.cc
@@ -46,6 +46,11 @@ network::mojom::blink::ReferrerPolicy PolicyContainer::GetReferrerPolicy()
   return policies_->referrer_policy;
 }
 
+network::mojom::blink::IPAddressSpace PolicyContainer::GetIPAddressSpace()
+    const {
+  return policies_->ip_address_space;
+}
+
 void PolicyContainer::UpdateReferrerPolicy(
     network::mojom::blink::ReferrerPolicy policy,
     const base::UnguessableToken& initiator_state_token) {
@@ -55,6 +60,11 @@ void PolicyContainer::UpdateReferrerPolicy(
                                                    initiator_state_token);
 }
 
+void PolicyContainer::SetIPAddressSpace(
+    network::mojom::IPAddressSpace ip_address_space) {
+  policies_->ip_address_space = ip_address_space;
+}
+
 const mojom::blink::PolicyContainerPolicies& PolicyContainer::GetPolicies()
     const {
   return *policies_;
diff --git a/third_party/blink/renderer/core/frame/policy_container.h b/third_party/blink/renderer/core/frame/policy_container.h
index ea097c9dc669fe5e4d204eded8f1b7e736d96fc6..1d99c3edd9e5b54fc4f1c76239fdbf522be74f04 100644
--- a/third_party/blink/renderer/core/frame/policy_container.h
+++ b/third_party/blink/renderer/core/frame/policy_container.h
@@ -8,6 +8,7 @@
 #include "base/unguessable_token.h"
 #include "mojo/public/cpp/bindings/associated_remote.h"
 #include "services/network/public/mojom/content_security_policy.mojom-blink-forward.h"
+#include "services/network/public/mojom/ip_address_space.mojom-shared.h"
 #include "services/network/public/mojom/referrer_policy.mojom-shared.h"
 #include "services/network/public/mojom/web_sandbox_flags.mojom-shared.h"
 #include "third_party/blink/public/mojom/frame/policy_container.mojom-blink.h"
@@ -49,6 +50,16 @@ class CORE_EXPORT PolicyContainer {
       const base::UnguessableToken& new_initiator_state_token);
   network::mojom::blink::ReferrerPolicy GetReferrerPolicy() const;
 
+  // This setter is used only by worklets and workers, which do not sync the
+  // PolicyContainer with the browser.
+  //
+  // TODO(https://crbug.com/1177199): Remove this when we implement policy
+  // inheritance for workers/worklets using the PolicyContainer.
+  void SetIPAddressSpace(
+      network::mojom::IPAddressSpace ip_address_space);
+
+  network::mojom::blink::IPAddressSpace GetIPAddressSpace() const;
+
   // Append |policies| to the list of Content Security Policy and sync them with
   // the PolicyContainerHost.
   // A new initiator state token should be generated and passed to this
diff --git a/third_party/blink/renderer/core/layout/custom/layout_worklet_global_scope_proxy.cc b/third_party/blink/renderer/core/layout/custom/layout_worklet_global_scope_proxy.cc
index 450a8d27e306c5eda9bede5f007b83ce645e46a7..b0923517d6737eb448a2d457dc2b196b4784189b 100644
--- a/third_party/blink/renderer/core/layout/custom/layout_worklet_global_scope_proxy.cc
+++ b/third_party/blink/renderer/core/layout/custom/layout_worklet_global_scope_proxy.cc
@@ -49,6 +49,7 @@ LayoutWorkletGlobalScopeProxy::LayoutWorkletGlobalScopeProxy(
       window->GetSecurityOrigin(), window->IsSecureContext(),
       window->GetHttpsState(), nullptr /* worker_clients */,
       frame_client->CreateWorkerContentSettingsClient(),
+      window->AddressSpace(),
       OriginTrialContext::GetInheritedTrialFeatures(window).get(),
       base::UnguessableToken::Create(), nullptr /* worker_settings */,
       mojom::blink::V8CacheOptions::kDefault, module_responses_map,
diff --git a/third_party/blink/renderer/core/loader/modulescript/installed_service_worker_module_script_fetcher.cc b/third_party/blink/renderer/core/loader/modulescript/installed_service_worker_module_script_fetcher.cc
index 3bcd8a2b8b6194191f4b1933a6d99182343845ce..a5a5af66fd9b6afc1dc9068360e2dd6a4daafa61 100644
--- a/third_party/blink/renderer/core/loader/modulescript/installed_service_worker_module_script_fetcher.cc
+++ b/third_party/blink/renderer/core/loader/modulescript/installed_service_worker_module_script_fetcher.cc
@@ -70,6 +70,7 @@ void InstalledServiceWorkerModuleScriptFetcher::Fetch(
 
     global_scope_->Initialize(
         response_url, response_referrer_policy,
+        script_data->GetResponseAddressSpace(),
         ParseContentSecurityPolicyHeaders(
             script_data->GetContentSecurityPolicyResponseHeaders()),
         // TODO(crbug.com/488089240): Plumb Document Policy in Service Workers.
diff --git a/third_party/blink/renderer/core/loader/modulescript/module_script_loader_test.cc b/third_party/blink/renderer/core/loader/modulescript/module_script_loader_test.cc
index feb72cca552665b19c42d4a44c6df0b1c5ca2d1e..3ad1530df4c14363b1c105ed866aa8655babb30c 100644
--- a/third_party/blink/renderer/core/loader/modulescript/module_script_loader_test.cc
+++ b/third_party/blink/renderer/core/loader/modulescript/module_script_loader_test.cc
@@ -227,6 +227,7 @@ void ModuleScriptLoaderTest::InitializeForWorklet() {
       DocumentPolicy::DocumentPolicyBundle{}, security_origin_.get(),
       true /* is_secure_context */, HttpsState::kModern,
       nullptr /* worker_clients */, nullptr /* content_settings_client */,
+      network::mojom::IPAddressSpace::kLocal,
       nullptr /* inherited_trial_features */, base::UnguessableToken::Create(),
       nullptr /* worker_settings */, mojom::blink::V8CacheOptions::kDefault,
       MakeGarbageCollected<WorkletModuleResponsesMap>(),
diff --git a/third_party/blink/renderer/core/loader/modulescript/worker_module_script_fetcher.cc b/third_party/blink/renderer/core/loader/modulescript/worker_module_script_fetcher.cc
index 0403d706ad330c454aa5858e9dc0b61139895789..ab534d2f56087a43a7e2405cccb6247b356c7f4d 100644
--- a/third_party/blink/renderer/core/loader/modulescript/worker_module_script_fetcher.cc
+++ b/third_party/blink/renderer/core/loader/modulescript/worker_module_script_fetcher.cc
@@ -6,6 +6,7 @@
 
 #include <memory>
 
+#include "services/network/public/mojom/ip_address_space.mojom-blink.h"
 #include "services/network/public/mojom/referrer_policy.mojom-blink.h"
 #include "third_party/blink/public/common/features.h"
 #include "third_party/blink/public/common/loader/network_utils.h"
@@ -192,7 +193,7 @@ void WorkerModuleScriptFetcher::NotifyClient(
 
     // Step 12.3-12.6 are implemented in Initialize().
     global_scope_->Initialize(
-        response_url, response_referrer_policy,
+        response_url, response_referrer_policy, response.AddressSpace(),
         ParseContentSecurityPolicyHeaders(
             ContentSecurityPolicyResponseHeaders(response)),
         // TODO(crbug.com/488090079): Plumb Document Policy in Module Workers.
diff --git a/third_party/blink/renderer/core/workers/dedicated_worker.cc b/third_party/blink/renderer/core/workers/dedicated_worker.cc
index bc8726d68af801c093bc4baf266c7d3271638df8..d2454beb43f60cb8037a3db950e38ebe85b77d57 100644
--- a/third_party/blink/renderer/core/workers/dedicated_worker.cc
+++ b/third_party/blink/renderer/core/workers/dedicated_worker.cc
@@ -368,6 +368,7 @@ void DedicatedWorker::OnScriptLoadStarted(
                 network::mojom::ReferrerPolicy::kDefault,
                 Vector<network::mojom::blink::ContentSecurityPolicyPtr>(),
                 DocumentPolicy::DocumentPolicyBundle{},
+                std::nullopt /* response_address_space */,
                 std::move(back_forward_cache_controller_host),
                 std::move(policy_container), std::move(coep_reporting_observer),
                 std::move(dip_reporting_observer));
@@ -409,6 +410,7 @@ void DedicatedWorker::ContinueStart(
     Vector<network::mojom::blink::ContentSecurityPolicyPtr>
         response_content_security_policies,
     DocumentPolicy::DocumentPolicyBundle response_document_policy,
+    std::optional<network::mojom::IPAddressSpace> response_address_space,
     mojo::PendingRemote<mojom::blink::BackForwardCacheControllerHost>
         back_forward_cache_controller_host,
     std::unique_ptr<WebPolicyContainer> policy_container,
@@ -435,6 +437,7 @@ void DedicatedWorker::ContinueStart(
                      std::move(referrer_policy),
                      std::move(response_content_security_policies),
                      std::move(response_document_policy),
+                     response_address_space,
                      std::move(back_forward_cache_controller_host),
                      std::move(policy_container),
                      std::move(coep_reporting_observer),
@@ -446,6 +449,7 @@ void DedicatedWorker::ContinueStart(
       script_url, std::move(worker_main_script_load_params),
       std::move(referrer_policy), std::move(response_content_security_policies),
       std::move(response_document_policy),
+      response_address_space,
       std::move(back_forward_cache_controller_host),
       std::move(policy_container), std::move(coep_reporting_observer),
       std::move(dip_reporting_observer));
@@ -459,6 +463,7 @@ void DedicatedWorker::ContinueStartInternal(
     Vector<network::mojom::blink::ContentSecurityPolicyPtr>
         response_content_security_policies,
     DocumentPolicy::DocumentPolicyBundle document_policy,
+    std::optional<network::mojom::IPAddressSpace> response_address_space,
     mojo::PendingRemote<mojom::blink::BackForwardCacheControllerHost>
         back_forward_cache_controller_host,
     std::unique_ptr<WebPolicyContainer> policy_container,
@@ -474,7 +479,7 @@ void DedicatedWorker::ContinueStartInternal(
       CreateGlobalScopeCreationParams(
           script_url, referrer_policy,
           std::move(response_content_security_policies),
-          std::move(document_policy), std::move(coep_reporting_observer),
+          std::move(document_policy), response_address_space, std::move(coep_reporting_observer),
           std::move(dip_reporting_observer)),
       std::move(worker_main_script_load_params), options_, script_url,
       *outside_fetch_client_settings_object_, v8_stack_trace_id_, token_,
@@ -514,6 +519,7 @@ DedicatedWorker::CreateGlobalScopeCreationParams(
     Vector<network::mojom::blink::ContentSecurityPolicyPtr>
         response_content_security_policies,
     DocumentPolicy::DocumentPolicyBundle document_policy,
+    std::optional<network::mojom::IPAddressSpace> response_address_space,
     mojo::PendingReceiver<mojom::blink::ReportingObserver>
         coep_reporting_observer,
     mojo::PendingReceiver<mojom::blink::ReportingObserver>
@@ -566,6 +572,7 @@ DedicatedWorker::CreateGlobalScopeCreationParams(
       std::move(document_policy), execution_context->GetSecurityOrigin(),
       execution_context->IsSecureContext(), execution_context->GetHttpsState(),
       MakeGarbageCollected<WorkerClients>(), CreateWebContentSettingsClient(),
+      response_address_space,
       OriginTrialContext::GetInheritedTrialFeatures(execution_context).get(),
       parent_devtools_token, std::move(settings),
       mojom::blink::V8CacheOptions::kDefault,
diff --git a/third_party/blink/renderer/core/workers/dedicated_worker.h b/third_party/blink/renderer/core/workers/dedicated_worker.h
index 7253aa186039a0800673a1d0fdf97a80c7d6dc2a..f02d5610505edf3f4bfc8f33b5a92d06fa3aae63 100644
--- a/third_party/blink/renderer/core/workers/dedicated_worker.h
+++ b/third_party/blink/renderer/core/workers/dedicated_worker.h
@@ -161,6 +161,7 @@ class CORE_EXPORT DedicatedWorker final
       Vector<network::mojom::blink::ContentSecurityPolicyPtr>
           response_content_security_policies,
       DocumentPolicy::DocumentPolicyBundle response_document_policy,
+      std::optional<network::mojom::IPAddressSpace> response_address_space,
       mojo::PendingRemote<mojom::blink::BackForwardCacheControllerHost>
           back_forward_cache_controller_host,
       std::unique_ptr<WebPolicyContainer> policy_container,
@@ -176,6 +177,7 @@ class CORE_EXPORT DedicatedWorker final
       Vector<network::mojom::blink::ContentSecurityPolicyPtr>
           response_content_security_policies,
       DocumentPolicy::DocumentPolicyBundle response_document_policy,
+      std::optional<network::mojom::IPAddressSpace> response_address_space,
       mojo::PendingRemote<mojom::blink::BackForwardCacheControllerHost>
           back_forward_cache_controller_host,
       std::unique_ptr<WebPolicyContainer> policy_container,
@@ -189,6 +191,7 @@ class CORE_EXPORT DedicatedWorker final
       Vector<network::mojom::blink::ContentSecurityPolicyPtr>
           response_content_security_policies,
       DocumentPolicy::DocumentPolicyBundle response_document_policy,
+      std::optional<network::mojom::IPAddressSpace> response_address_space,
       mojo::PendingReceiver<mojom::blink::ReportingObserver>
           coep_reporting_observer,
       mojo::PendingReceiver<mojom::blink::ReportingObserver>
diff --git a/third_party/blink/renderer/core/workers/dedicated_worker_global_scope.cc b/third_party/blink/renderer/core/workers/dedicated_worker_global_scope.cc
index 9242721a602fca21de75f8352c565b99b91f85cb..3f792c4567869ade7e7c196425de1fa566d488f0 100644
--- a/third_party/blink/renderer/core/workers/dedicated_worker_global_scope.cc
+++ b/third_party/blink/renderer/core/workers/dedicated_worker_global_scope.cc
@@ -118,8 +118,10 @@ DedicatedWorkerGlobalScope* DedicatedWorkerGlobalScope::Create(
     // Legacy on-the-main-thread worker script fetch (to be removed):
     // Pass dummy origin trial tokens here as it is already set to outside's
     // origin trial tokens in DedicatedWorkerGlobalScope's constructor.
+    std::optional<network::mojom::IPAddressSpace> response_address_space =
+        creation_params->response_address_space;
     global_scope->Initialize(response_script_url, response_referrer_policy,
-                             std::move(response_csp),
+                             *response_address_space, std::move(response_csp),
                              std::move(response_document_policy),
                              nullptr /* response_origin_trial_tokens */);
     return global_scope;
@@ -241,6 +243,7 @@ const AtomicString& DedicatedWorkerGlobalScope::InterfaceName() const {
 void DedicatedWorkerGlobalScope::Initialize(
     const KURL& response_url,
     network::mojom::ReferrerPolicy response_referrer_policy,
+    network::mojom::IPAddressSpace response_address_space,
     Vector<network::mojom::blink::ContentSecurityPolicyPtr> response_csp,
     DocumentPolicy::DocumentPolicyBundle response_document_policy,
     const Vector<String>* /* response_origin_trial_tokens */) {
@@ -257,6 +260,9 @@ void DedicatedWorkerGlobalScope::Initialize(
   // parsing the `Referrer-Policy` header of response."
   SetReferrerPolicy(response_referrer_policy);
 
+  // https://wicg.github.io/cors-rfc1918/#integration-html
+  SetAddressSpace(response_address_space);
+
   // The following is the Content-Security-Policy part of "Initialize worker
   // global scope's policy container"
   // https://html.spec.whatwg.org/#initialize-worker-policy-container
@@ -515,6 +521,7 @@ void DedicatedWorkerGlobalScope::DidFetchClassicScript(
   // Pass dummy origin trial tokens here as it is already set to outside's
   // origin trial tokens in DedicatedWorkerGlobalScope's constructor.
   Initialize(classic_script_loader->ResponseURL(), response_referrer_policy,
+             classic_script_loader->ResponseAddressSpace(),
              classic_script_loader->GetContentSecurityPolicy()
                  ? mojo::Clone(classic_script_loader->GetContentSecurityPolicy()
                                    ->GetParsedPolicies())
diff --git a/third_party/blink/renderer/core/workers/dedicated_worker_global_scope.h b/third_party/blink/renderer/core/workers/dedicated_worker_global_scope.h
index ee82e8c8c3654e8715db69e2f2ef8f9b778dd4d6..dc3ae3d59b1aefd41fb17281e45b74ad6dd9e3d7 100644
--- a/third_party/blink/renderer/core/workers/dedicated_worker_global_scope.h
+++ b/third_party/blink/renderer/core/workers/dedicated_worker_global_scope.h
@@ -111,6 +111,7 @@ class CORE_EXPORT DedicatedWorkerGlobalScope final : public WorkerGlobalScope {
   void Initialize(
       const KURL& response_url,
       network::mojom::ReferrerPolicy response_referrer_policy,
+      network::mojom::IPAddressSpace response_address_space,
       Vector<network::mojom::blink::ContentSecurityPolicyPtr> response_csp,
       DocumentPolicy::DocumentPolicyBundle response_document_policy,
       const Vector<String>* response_origin_trial_tokens) override;
diff --git a/third_party/blink/renderer/core/workers/global_scope_creation_params.cc b/third_party/blink/renderer/core/workers/global_scope_creation_params.cc
index 2ce7145dc63f2904d01076d99a24ba05a91dc182..8eb242c87130da0266807f89b1e30f1a8fa08f19 100644
--- a/third_party/blink/renderer/core/workers/global_scope_creation_params.cc
+++ b/third_party/blink/renderer/core/workers/global_scope_creation_params.cc
@@ -35,6 +35,7 @@ GlobalScopeCreationParams::GlobalScopeCreationParams(
     HttpsState starter_https_state,
     WorkerClients* worker_clients,
     std::unique_ptr<WebContentSettingsClient> content_settings_client,
+    std::optional<network::mojom::IPAddressSpace> response_address_space,
     const Vector<mojom::blink::OriginTrialFeature>* inherited_trial_features,
     const base::UnguessableToken& parent_devtools_token,
     std::unique_ptr<WorkerSettings> worker_settings,
@@ -81,6 +82,7 @@ GlobalScopeCreationParams::GlobalScopeCreationParams(
       starter_https_state(starter_https_state),
       worker_clients(worker_clients),
       content_settings_client(std::move(content_settings_client)),
+      response_address_space(response_address_space),
       parent_devtools_token(parent_devtools_token),
       worker_settings(std::move(worker_settings)),
       v8_cache_options(v8_cache_options),
@@ -148,6 +150,7 @@ GlobalScopeCreationParams::CreateForWorkerForTesting(
       /*starter_secure_context=*/false, CalculateHttpsState(starter_origin),
       MakeGarbageCollected<WorkerClients>(),
       /*content_settings_client=*/nullptr,
+      network::mojom::IPAddressSpace::kLocal,
       /*inherited_trial_features=*/nullptr, base::UnguessableToken::Create(),
       std::move(worker_settings), mojom::blink::V8CacheOptions::kDefault,
       /*worklet_module_responses_map=*/nullptr, mojo::NullRemote(),
diff --git a/third_party/blink/renderer/core/workers/global_scope_creation_params.h b/third_party/blink/renderer/core/workers/global_scope_creation_params.h
index f7f92a38ff6e4fac76dec40b62131af36c3675f4..9ff9d513dfa207582458d03cae026ad7c07e6b60 100644
--- a/third_party/blink/renderer/core/workers/global_scope_creation_params.h
+++ b/third_party/blink/renderer/core/workers/global_scope_creation_params.h
@@ -14,6 +14,7 @@
 #include "net/storage_access_api/status.h"
 #include "services/metrics/public/cpp/ukm_source_id.h"
 #include "services/network/public/cpp/permissions_policy/permissions_policy.h"
+#include "services/network/public/mojom/ip_address_space.mojom-blink-forward.h"
 #include "services/network/public/mojom/referrer_policy.mojom-blink-forward.h"
 #include "third_party/blink/public/common/permissions_policy/document_policy.h"
 #include "third_party/blink/public/common/tokens/tokens.h"
@@ -65,6 +66,7 @@ struct CORE_EXPORT GlobalScopeCreationParams final {
       HttpsState starter_https_state,
       WorkerClients*,
       std::unique_ptr<WebContentSettingsClient>,
+      std::optional<network::mojom::IPAddressSpace>,
       const Vector<mojom::blink::OriginTrialFeature>* inherited_trial_features,
       const base::UnguessableToken& parent_devtools_token,
       std::unique_ptr<WorkerSettings>,
@@ -200,6 +202,11 @@ struct CORE_EXPORT GlobalScopeCreationParams final {
 
   std::unique_ptr<WebContentSettingsClient> content_settings_client;
 
+  // Worker script response's address space. This is valid only when the worker
+  // script is fetched on the main thread (i.e., when
+  // |off_main_thread_fetch_option| is kDisabled).
+  std::optional<network::mojom::IPAddressSpace> response_address_space;
+
   base::UnguessableToken parent_devtools_token;
 
   std::unique_ptr<WorkerSettings> worker_settings;
diff --git a/third_party/blink/renderer/core/workers/installed_scripts_manager.cc b/third_party/blink/renderer/core/workers/installed_scripts_manager.cc
index fb66bdda0ea886eb0dab4bd24f55cb8ff8108bf0..2a978cb476740110eb75b61a38b74ad0edb60cdd 100644
--- a/third_party/blink/renderer/core/workers/installed_scripts_manager.cc
+++ b/third_party/blink/renderer/core/workers/installed_scripts_manager.cc
@@ -4,6 +4,7 @@
 
 #include "third_party/blink/renderer/core/workers/installed_scripts_manager.h"
 
+#include "services/network/public/mojom/ip_address_space.mojom-blink.h"
 #include "third_party/blink/renderer/core/frame/csp/content_security_policy.h"
 #include "third_party/blink/renderer/core/origin_trials/origin_trial_context.h"
 #include "third_party/blink/renderer/platform/network/http_names.h"
@@ -20,6 +21,21 @@ InstalledScriptsManager::ScriptData::ScriptData(
       source_text_(std::move(source_text)),
       meta_data_(std::move(meta_data)) {
   headers_.Adopt(std::move(header_data));
+
+  // Calculate an address space from worker script's response url according to
+  // the "CORS and RFC1918" spec:
+  // https://wicg.github.io/cors-rfc1918/#integration-html
+  //
+  // Currently this implementation is not fully consistent with the spec for
+  // historical reasons.
+  // TODO(https://crbug.com/955213): Make this consistent with the spec.
+  // TODO(https://crbug.com/955213): Move this function to a more appropriate
+  // place so that this is shareable out of worker code.
+  response_address_space_ = network::mojom::IPAddressSpace::kPublic;
+  if (network_utils::IsReservedIPAddress(script_url_.Host()))
+    response_address_space_ = network::mojom::IPAddressSpace::kLocal;
+  if (SecurityOrigin::Create(script_url_)->IsLocalhost())
+    response_address_space_ = network::mojom::IPAddressSpace::kLoopback;
 }
 
 ContentSecurityPolicyResponseHeaders
diff --git a/third_party/blink/renderer/core/workers/installed_scripts_manager.h b/third_party/blink/renderer/core/workers/installed_scripts_manager.h
index da26ae8b39d8f025d462e2e2a509dd90de7fe9ea..9be2e6151162f89b92cc6a3646446194dc75afe6 100644
--- a/third_party/blink/renderer/core/workers/installed_scripts_manager.h
+++ b/third_party/blink/renderer/core/workers/installed_scripts_manager.h
@@ -5,6 +5,7 @@
 #ifndef THIRD_PARTY_BLINK_RENDERER_CORE_WORKERS_INSTALLED_SCRIPTS_MANAGER_H_
 #define THIRD_PARTY_BLINK_RENDERER_CORE_WORKERS_INSTALLED_SCRIPTS_MANAGER_H_
 
+#include "services/network/public/mojom/ip_address_space.mojom-blink-forward.h"
 #include "third_party/blink/renderer/core/core_export.h"
 #include "third_party/blink/renderer/platform/network/content_security_policy_response_headers.h"
 #include "third_party/blink/renderer/platform/network/http_header_map.h"
@@ -45,6 +46,9 @@ class InstalledScriptsManager {
     GetContentSecurityPolicyResponseHeaders();
     String GetReferrerPolicy();
     String GetHttpContentType();
+    network::mojom::IPAddressSpace GetResponseAddressSpace() const {
+      return response_address_space_;
+    }
     std::unique_ptr<Vector<String>> CreateOriginTrialTokens();
 
    private:
@@ -52,6 +56,7 @@ class InstalledScriptsManager {
     String source_text_;
     std::unique_ptr<Vector<uint8_t>> meta_data_;
     HTTPHeaderMap headers_;
+    network::mojom::IPAddressSpace response_address_space_;
   };
 
   // Used on the main or worker thread. Returns true if the script has been
diff --git a/third_party/blink/renderer/core/workers/main_thread_worklet_test.cc b/third_party/blink/renderer/core/workers/main_thread_worklet_test.cc
index ce7baa0f7fbf947751c6110a6dab7fd12817d820..09d341193f6714db65c33498966a86c69c178e11 100644
--- a/third_party/blink/renderer/core/workers/main_thread_worklet_test.cc
+++ b/third_party/blink/renderer/core/workers/main_thread_worklet_test.cc
@@ -73,6 +73,7 @@ class MainThreadWorkletTest : public PageTestBase {
         window->GetSecurityOrigin(), window->IsSecureContext(),
         window->GetHttpsState(), nullptr /* worker_clients */,
         nullptr /* content_settings_client */,
+        window->AddressSpace(),
         OriginTrialContext::GetInheritedTrialFeatures(window).get(),
         base::UnguessableToken::Create(), nullptr /* worker_settings */,
         mojom::blink::V8CacheOptions::kDefault,
diff --git a/third_party/blink/renderer/core/workers/shared_worker_global_scope.cc b/third_party/blink/renderer/core/workers/shared_worker_global_scope.cc
index a458d2a79a8d92a2194a7285a105446d13e4e894..7192f887ac9200e2a220ed69c265b0324d6f20bd 100644
--- a/third_party/blink/renderer/core/workers/shared_worker_global_scope.cc
+++ b/third_party/blink/renderer/core/workers/shared_worker_global_scope.cc
@@ -82,6 +82,7 @@ const AtomicString& SharedWorkerGlobalScope::InterfaceName() const {
 void SharedWorkerGlobalScope::Initialize(
     const KURL& response_url,
     network::mojom::ReferrerPolicy response_referrer_policy,
+    network::mojom::IPAddressSpace response_address_space,
     Vector<network::mojom::blink::ContentSecurityPolicyPtr> response_csp,
     DocumentPolicy::DocumentPolicyBundle response_document_policy,
     const Vector<String>* response_origin_trial_tokens) {
@@ -96,6 +97,9 @@ void SharedWorkerGlobalScope::Initialize(
   // parsing the `Referrer-Policy` header of response."
   SetReferrerPolicy(response_referrer_policy);
 
+  // https://wicg.github.io/cors-rfc1918/#integration-html
+  SetAddressSpace(response_address_space);
+
   // Step 12.6. "Execute the Initialize a global object's CSP list algorithm
   // on worker global scope and response. [CSP]"
   // SharedWorkerGlobalScope inherits the outside's CSP instead of the response
@@ -105,6 +109,8 @@ void SharedWorkerGlobalScope::Initialize(
   // https://fetch.spec.whatwg.org/#local-scheme
   //
   // https://w3c.github.io/webappsec-csp/#initialize-global-object-csp
+  // These should be called after SetAddressSpace() to correctly override the
+  // address space by the "treat-as-public-address" CSP directive.
   Vector<network::mojom::blink::ContentSecurityPolicyPtr> csp_headers =
       response_url.ProtocolIsAbout() || response_url.ProtocolIsData() ||
               response_url.ProtocolIs("blob")
@@ -251,6 +257,7 @@ void SharedWorkerGlobalScope::DidFetchClassicScript(
 
   // Step 12.3-12.6 are implemented in Initialize().
   Initialize(classic_script_loader->ResponseURL(), response_referrer_policy,
+             classic_script_loader->ResponseAddressSpace(),
              classic_script_loader->GetContentSecurityPolicy()
                  ? mojo::Clone(classic_script_loader->GetContentSecurityPolicy()
                                    ->GetParsedPolicies())
diff --git a/third_party/blink/renderer/core/workers/shared_worker_global_scope.h b/third_party/blink/renderer/core/workers/shared_worker_global_scope.h
index 1a52719f7f74554b07a7692ff489f34cb3f5b037..3a1478b0e88434c4de79a076eb7c5d24c0da0b23 100644
--- a/third_party/blink/renderer/core/workers/shared_worker_global_scope.h
+++ b/third_party/blink/renderer/core/workers/shared_worker_global_scope.h
@@ -65,6 +65,7 @@ class CORE_EXPORT SharedWorkerGlobalScope final : public WorkerGlobalScope {
   void Initialize(
       const KURL& response_url,
       network::mojom::ReferrerPolicy response_referrer_policy,
+      network::mojom::IPAddressSpace response_address_space,
       Vector<network::mojom::blink::ContentSecurityPolicyPtr> response_csp,
       DocumentPolicy::DocumentPolicyBundle response_document_policy,
       const Vector<String>* response_origin_trial_tokens) override;
diff --git a/third_party/blink/renderer/core/workers/threaded_worklet_messaging_proxy.cc b/third_party/blink/renderer/core/workers/threaded_worklet_messaging_proxy.cc
index 3cf2c103c9460d6a5ef64b946a57f96432501670..aac3bd4fcf9e5dacf6e57164eaa13eab810b9d33 100644
--- a/third_party/blink/renderer/core/workers/threaded_worklet_messaging_proxy.cc
+++ b/third_party/blink/renderer/core/workers/threaded_worklet_messaging_proxy.cc
@@ -94,7 +94,9 @@ void ThreadedWorkletMessagingProxy::Initialize(
         starter_secure_context,
         /*starter_https_state=*/HttpsState::kNone,
         /*worker_clients=*/nullptr,
-        /*content_settings_client=*/nullptr, &inherited_trial_features,
+        /*content_settings_client=*/nullptr,
+        /*response_address_space*/network::mojom::IPAddressSpace::kUnknown,
+        &inherited_trial_features,
         /*parent_devtools_token=*/
         client_provided_global_scope_creation_params->devtools_token,
         /*worker_settings=*/nullptr,
@@ -146,6 +148,7 @@ void ThreadedWorkletMessagingProxy::Initialize(
           window->GetSecurityOrigin(), window->IsSecureContext(),
           window->GetHttpsState(), worker_clients,
           frame_client->CreateWorkerContentSettingsClient(),
+          window->AddressSpace(),
           OriginTrialContext::GetInheritedTrialFeatures(window).get(),
           base::UnguessableToken::Create(),
           std::make_unique<WorkerSettings>(window->GetFrame()->GetSettings()),
diff --git a/third_party/blink/renderer/core/workers/threaded_worklet_test.cc b/third_party/blink/renderer/core/workers/threaded_worklet_test.cc
index 968e7e2bbf413532c75e0dd0220387199ec5f337..296856d2db2aecb13774eed71dbe53b8df16d52c 100644
--- a/third_party/blink/renderer/core/workers/threaded_worklet_test.cc
+++ b/third_party/blink/renderer/core/workers/threaded_worklet_test.cc
@@ -220,6 +220,7 @@ class ThreadedWorkletMessagingProxyForTest
             GetExecutionContext()->IsSecureContext(),
             GetExecutionContext()->GetHttpsState(), worker_clients,
             nullptr /* content_settings_client */,
+            GetExecutionContext()->AddressSpace(),
             OriginTrialContext::GetInheritedTrialFeatures(GetExecutionContext())
                 .get(),
             base::UnguessableToken::Create(), std::move(worker_settings),
diff --git a/third_party/blink/renderer/core/workers/worker_classic_script_loader.cc b/third_party/blink/renderer/core/workers/worker_classic_script_loader.cc
index 183068ded5b8c709211f3bd6865b19592538840f..760297cccc2f000a1d15ace282232ed90396c80e 100644
--- a/third_party/blink/renderer/core/workers/worker_classic_script_loader.cc
+++ b/third_party/blink/renderer/core/workers/worker_classic_script_loader.cc
@@ -30,6 +30,8 @@
 #include <memory>
 
 #include "base/memory/scoped_refptr.h"
+#include "services/network/public/mojom/ip_address_space.mojom-blink.h"
+#include "third_party/blink/public/common/features.h"
 #include "third_party/blink/public/mojom/fetch/fetch_api_request.mojom-blink.h"
 #include "third_party/blink/renderer/core/execution_context/execution_context.h"
 #include "third_party/blink/renderer/core/frame/csp/content_security_policy.h"
@@ -101,7 +103,8 @@ String CheckSameOriginEnforcement(const KURL& request_url,
 
 }  // namespace
 
-WorkerClassicScriptLoader::WorkerClassicScriptLoader() {}
+WorkerClassicScriptLoader::WorkerClassicScriptLoader()
+    : response_address_space_(network::mojom::IPAddressSpace::kPublic) {}
 
 void WorkerClassicScriptLoader::LoadSynchronously(
     ExecutionContext& execution_context,
@@ -235,6 +238,7 @@ void WorkerClassicScriptLoader::DidReceiveResponse(
   identifier_ = identifier;
   response_url_ = response.ResponseUrl();
   response_encoding_ = response.TextEncodingName();
+  response_address_space_ = response.AddressSpace();
 
   referrer_policy_ = response.HttpHeaderField(http_names::kReferrerPolicy);
   ProcessContentSecurityPolicy(response);
diff --git a/third_party/blink/renderer/core/workers/worker_classic_script_loader.h b/third_party/blink/renderer/core/workers/worker_classic_script_loader.h
index 6d4d8a072eed2f0495dc22abfdb7ba6e543a1032..e2c4f11873f9e91c42235770f1f8829663d64a45 100644
--- a/third_party/blink/renderer/core/workers/worker_classic_script_loader.h
+++ b/third_party/blink/renderer/core/workers/worker_classic_script_loader.h
@@ -32,6 +32,7 @@
 
 #include "mojo/public/cpp/bindings/pending_remote.h"
 #include "services/network/public/mojom/fetch_api.mojom-blink-forward.h"
+#include "services/network/public/mojom/ip_address_space.mojom-blink-forward.h"
 #include "services/network/public/mojom/url_loader_factory.mojom-blink.h"
 #include "third_party/blink/public/common/permissions_policy/document_policy.h"
 #include "third_party/blink/public/mojom/loader/resource_load_info_notifier.mojom-shared.h"
@@ -124,6 +125,10 @@ class CORE_EXPORT WorkerClassicScriptLoader final
     return document_policy_;
   }
 
+  network::mojom::IPAddressSpace ResponseAddressSpace() const {
+    return response_address_space_;
+  }
+
   const Vector<String>* OriginTrialTokens() const {
     return origin_trial_tokens_.get();
   }
@@ -179,6 +184,7 @@ class CORE_EXPORT WorkerClassicScriptLoader final
   uint64_t identifier_ = 0;
   std::unique_ptr<Vector<uint8_t>> cached_metadata_;
   Member<ContentSecurityPolicy> content_security_policy_;
+  network::mojom::IPAddressSpace response_address_space_;
   std::unique_ptr<Vector<String>> origin_trial_tokens_;
   String referrer_policy_;
   DocumentPolicy::DocumentPolicyBundle document_policy_;
diff --git a/third_party/blink/renderer/core/workers/worker_global_scope.h b/third_party/blink/renderer/core/workers/worker_global_scope.h
index 44d8d345f15cc54c509e965757819dc0ee9f841d..b291d064d34f2103063464adbb53cf158270e038 100644
--- a/third_party/blink/renderer/core/workers/worker_global_scope.h
+++ b/third_party/blink/renderer/core/workers/worker_global_scope.h
@@ -33,6 +33,7 @@
 #include "base/time/time.h"
 #include "services/network/public/mojom/fetch_api.mojom-blink-forward.h"
 #include "third_party/blink/public/common/permissions_policy/document_policy.h"
+#include "services/network/public/mojom/ip_address_space.mojom-blink-forward.h"
 #include "third_party/blink/public/common/tokens/tokens.h"
 #include "third_party/blink/public/mojom/loader/code_cache.mojom-blink-forward.h"
 #include "third_party/blink/public/mojom/script/script_type.mojom-blink-forward.h"
@@ -177,6 +178,7 @@ class CORE_EXPORT WorkerGlobalScope
   virtual void Initialize(
       const KURL& response_url,
       network::mojom::ReferrerPolicy response_referrer_policy,
+      network::mojom::IPAddressSpace response_address_space,
       Vector<network::mojom::blink::ContentSecurityPolicyPtr> response_csp,
       DocumentPolicy::DocumentPolicyBundle response_document_policy,
       const Vector<String>* response_origin_trial_tokens) = 0;
diff --git a/third_party/blink/renderer/core/workers/worker_thread_test.cc b/third_party/blink/renderer/core/workers/worker_thread_test.cc
index d332ae330747f94d2c533a290a524e85b75b8476..177210d6f365fdb6ceab8ba7d18d200020333990 100644
--- a/third_party/blink/renderer/core/workers/worker_thread_test.cc
+++ b/third_party/blink/renderer/core/workers/worker_thread_test.cc
@@ -10,6 +10,7 @@
 #include "base/run_loop.h"
 #include "base/synchronization/lock.h"
 #include "base/synchronization/waitable_event.h"
+#include "services/network/public/mojom/ip_address_space.mojom-blink.h"
 #include "testing/gmock/include/gmock/gmock.h"
 #include "testing/gtest/include/gtest/gtest.h"
 #include "third_party/blink/public/mojom/v8_cache_options.mojom-blink.h"
diff --git a/third_party/blink/renderer/core/workers/worker_thread_test_helper.h b/third_party/blink/renderer/core/workers/worker_thread_test_helper.h
index a0572c90e905741cda17f611997405494106e31e..97e6af5720f4a8507c3a16a5e408f03e3180d31d 100644
--- a/third_party/blink/renderer/core/workers/worker_thread_test_helper.h
+++ b/third_party/blink/renderer/core/workers/worker_thread_test_helper.h
@@ -10,6 +10,7 @@
 #include "base/synchronization/waitable_event.h"
 #include "mojo/public/cpp/bindings/receiver_set.h"
 #include "services/metrics/public/cpp/ukm_source_id.h"
+#include "services/network/public/mojom/ip_address_space.mojom-blink.h"
 #include "testing/gmock/include/gmock/gmock.h"
 #include "third_party/blink/public/common/loader/javascript_framework_detection.h"
 #include "third_party/blink/public/common/loader/worker_main_script_load_parameters.h"
@@ -80,12 +81,16 @@ class FakeWorkerGlobalScope : public WorkerGlobalScope {
   void Initialize(
       const KURL& response_url,
       network::mojom::ReferrerPolicy response_referrer_policy,
+      network::mojom::IPAddressSpace response_address_space,
       Vector<network::mojom::blink::ContentSecurityPolicyPtr> response_csp,
       DocumentPolicy::DocumentPolicyBundle response_document_policy,
       const Vector<String>* response_origin_trial_tokens) override {
     InitializeURL(response_url);
     SetReferrerPolicy(response_referrer_policy);
+    SetAddressSpace(response_address_space);
 
+    // These should be called after SetAddressSpace() to correctly override the
+    // address space by the "treat-as-public-address" CSP directive.
     InitContentSecurityPolicyFromVector(std::move(response_csp));
     BindContentSecurityPolicyToExecutionContext();
 
diff --git a/third_party/blink/renderer/modules/csspaint/paint_worklet_global_scope_proxy.cc b/third_party/blink/renderer/modules/csspaint/paint_worklet_global_scope_proxy.cc
index 6acafbe531f474389096b47b1408ab72836bbdcc..fd37f024ed7b0795aceef17622b0de8ad5b27c14 100644
--- a/third_party/blink/renderer/modules/csspaint/paint_worklet_global_scope_proxy.cc
+++ b/third_party/blink/renderer/modules/csspaint/paint_worklet_global_scope_proxy.cc
@@ -47,7 +47,7 @@ PaintWorkletGlobalScopeProxy::PaintWorkletGlobalScopeProxy(
       window->GetReferrerPolicy(), DocumentPolicy::DocumentPolicyBundle{},
       window->GetSecurityOrigin(), window->IsSecureContext(),
       window->GetHttpsState(), nullptr /* worker_clients */,
-      frame_client->CreateWorkerContentSettingsClient(),
+      frame_client->CreateWorkerContentSettingsClient(), window->AddressSpace(),
       OriginTrialContext::GetInheritedTrialFeatures(window).get(),
       base::UnguessableToken::Create(), nullptr /* worker_settings */,
       mojom::blink::V8CacheOptions::kDefault, module_responses_map,
diff --git a/third_party/blink/renderer/modules/exported/web_embedded_worker_impl.cc b/third_party/blink/renderer/modules/exported/web_embedded_worker_impl.cc
index e85f9c75e8ed6961ef75c1899e93ce12a9af3521..989f6097a5bab8555b1d7f525922c249115c7e78 100644
--- a/third_party/blink/renderer/modules/exported/web_embedded_worker_impl.cc
+++ b/third_party/blink/renderer/modules/exported/web_embedded_worker_impl.cc
@@ -224,6 +224,7 @@ void WebEmbeddedWorkerImpl::StartWorkerThread(
       DocumentPolicy::DocumentPolicyBundle{}, starter_origin.get(),
       starter_secure_context, starter_https_state,
       /*worker_clients=*/nullptr, std::move(content_settings_proxy),
+      std::nullopt /* response_address_space */,
       /*inherited_trial_features=*/nullptr,
       worker_start_data->devtools_worker_token, std::move(worker_settings),
   /*v8_cache_options=*/
diff --git a/third_party/blink/renderer/modules/service_worker/service_worker_global_scope.cc b/third_party/blink/renderer/modules/service_worker/service_worker_global_scope.cc
index 5dd32aae98bd12625cf6d0b68595bed3a4d40b71..0ca9cf7f8c1365aa63ff55dd96fe87a797e45135 100644
--- a/third_party/blink/renderer/modules/service_worker/service_worker_global_scope.cc
+++ b/third_party/blink/renderer/modules/service_worker/service_worker_global_scope.cc
@@ -470,6 +470,7 @@ void ServiceWorkerGlobalScope::DidFetchClassicScript(
   // is set, and with the following callback steps given evaluationStatus:"
   RunClassicScript(
       classic_script_loader->ResponseURL(), referrer_policy,
+      classic_script_loader->ResponseAddressSpace(),
       classic_script_loader->GetContentSecurityPolicy()
           ? mojo::Clone(classic_script_loader->GetContentSecurityPolicy()
                             ->GetParsedPolicies())
@@ -484,6 +485,7 @@ void ServiceWorkerGlobalScope::DidFetchClassicScript(
 void ServiceWorkerGlobalScope::Initialize(
     const KURL& response_url,
     network::mojom::ReferrerPolicy response_referrer_policy,
+    network::mojom::IPAddressSpace response_address_space,
     Vector<network::mojom::blink::ContentSecurityPolicyPtr> response_csp,
     DocumentPolicy::DocumentPolicyBundle response_document_policy,
     const Vector<String>* response_origin_trial_tokens) {
@@ -498,6 +500,9 @@ void ServiceWorkerGlobalScope::Initialize(
   // script resource's referrer policy."
   SetReferrerPolicy(response_referrer_policy);
 
+  // https://wicg.github.io/cors-rfc1918/#integration-html
+  SetAddressSpace(response_address_space);
+
   // This is quoted from the "Content Security Policy" algorithm in the service
   // workers spec:
   // "Whenever a user agent invokes Run Service Worker algorithm with a service
@@ -508,6 +513,9 @@ void ServiceWorkerGlobalScope::Initialize(
   // - If serviceWorker's script resource was delivered with a
   //   Content-Security-Policy-Report-Only HTTP header containing the value
   //   policy, the user agent must monitor policy for serviceWorker."
+  //
+  // These should be called after SetAddressSpace() to correctly override the
+  // address space by the "treat-as-public-address" CSP directive.
   InitContentSecurityPolicyFromVector(std::move(response_csp));
   BindContentSecurityPolicyToExecutionContext();
 
@@ -553,7 +561,7 @@ void ServiceWorkerGlobalScope::LoadAndRunInstalledClassicScript(
   }
 
   RunClassicScript(
-      script_url, referrer_policy,
+      script_url, referrer_policy, script_data->GetResponseAddressSpace(),
       ParseContentSecurityPolicyHeaders(
           script_data->GetContentSecurityPolicyResponseHeaders()),
       // TODO(crbug.com/488089240): Plumb Document Policy in Service Workers.
@@ -566,6 +574,7 @@ void ServiceWorkerGlobalScope::LoadAndRunInstalledClassicScript(
 void ServiceWorkerGlobalScope::RunClassicScript(
     const KURL& response_url,
     network::mojom::ReferrerPolicy response_referrer_policy,
+    network::mojom::IPAddressSpace response_address_space,
     Vector<network::mojom::blink::ContentSecurityPolicyPtr> response_csp,
     DocumentPolicy::DocumentPolicyBundle response_document_policy,
     const Vector<String>* response_origin_trial_tokens,
@@ -573,7 +582,7 @@ void ServiceWorkerGlobalScope::RunClassicScript(
     std::unique_ptr<Vector<uint8_t>> cached_meta_data,
     const v8_inspector::V8StackTraceId& stack_id) {
   // Step 4.5-4.11 are implemented in Initialize().
-  Initialize(response_url, response_referrer_policy, std::move(response_csp),
+  Initialize(response_url, response_referrer_policy, response_address_space, std::move(response_csp),
              std::move(response_document_policy), response_origin_trial_tokens);
 
   // Step 4.12. "Let evaluationStatus be the result of running the classic
diff --git a/third_party/blink/renderer/modules/service_worker/service_worker_global_scope.h b/third_party/blink/renderer/modules/service_worker/service_worker_global_scope.h
index d575f9b405980dd501624eb92dacc64e1e59b179..d68cc78f6ba10702884362bce93054c2cf4fe521 100644
--- a/third_party/blink/renderer/modules/service_worker/service_worker_global_scope.h
+++ b/third_party/blink/renderer/modules/service_worker/service_worker_global_scope.h
@@ -134,6 +134,7 @@ class MODULES_EXPORT ServiceWorkerGlobalScope final
   void Initialize(
       const KURL& response_url,
       network::mojom::ReferrerPolicy response_referrer_policy,
+      network::mojom::IPAddressSpace response_address_space,
       Vector<network::mojom::blink::ContentSecurityPolicyPtr> response_csp,
       DocumentPolicy::DocumentPolicyBundle response_document_policy,
       const Vector<String>* response_origin_trial_tokens) override;
@@ -426,6 +427,7 @@ class MODULES_EXPORT ServiceWorkerGlobalScope final
   void RunClassicScript(
       const KURL& response_url,
       network::mojom::ReferrerPolicy response_referrer_policy,
+      network::mojom::IPAddressSpace response_address_space,
       Vector<network::mojom::blink::ContentSecurityPolicyPtr> response_csp,
       DocumentPolicy::DocumentPolicyBundle response_document_policy,
       const Vector<String>* response_origin_trial_tokens,
diff --git a/third_party/blink/renderer/modules/webaudio/audio_worklet_global_scope_test.cc b/third_party/blink/renderer/modules/webaudio/audio_worklet_global_scope_test.cc
index 99b5306f4c658510447f6cbc67574c5c1ad47f6d..5c7a9399e55e8f0b0f69fb3386d797cb4814c346 100644
--- a/third_party/blink/renderer/modules/webaudio/audio_worklet_global_scope_test.cc
+++ b/third_party/blink/renderer/modules/webaudio/audio_worklet_global_scope_test.cc
@@ -92,6 +92,7 @@ class AudioWorkletGlobalScopeTest : public PageTestBase, public ModuleTestBase {
             window->GetSecurityOrigin(), window->IsSecureContext(),
             window->GetHttpsState(), nullptr /* worker_clients */,
             nullptr /* content_settings_client */,
+            window->AddressSpace(),
             OriginTrialContext::GetInheritedTrialFeatures(window).get(),
             base::UnguessableToken::Create(), nullptr /* worker_settings */,
             mojom::blink::V8CacheOptions::kDefault,
diff --git a/third_party/blink/renderer/modules/webaudio/audio_worklet_thread_test.cc b/third_party/blink/renderer/modules/webaudio/audio_worklet_thread_test.cc
index 569ad95ca3020c382720fb98c9d14a4d6f3dfb39..a589c66f41d2a7e6661dedff5fe67787b03e376c 100644
--- a/third_party/blink/renderer/modules/webaudio/audio_worklet_thread_test.cc
+++ b/third_party/blink/renderer/modules/webaudio/audio_worklet_thread_test.cc
@@ -109,6 +109,7 @@ class AudioWorkletThreadTest : public PageTestBase, public ModuleTestBase {
             window->GetSecurityOrigin(), window->IsSecureContext(),
             window->GetHttpsState(), nullptr /* worker_clients */,
             nullptr /* content_settings_client */,
+            window->AddressSpace(),
             OriginTrialContext::GetInheritedTrialFeatures(window).get(),
             base::UnguessableToken::Create(), nullptr /* worker_settings */,
             mojom::blink::V8CacheOptions::kDefault,
diff --git a/third_party/blink/renderer/modules/worklet/worklet_thread_test_common.cc b/third_party/blink/renderer/modules/worklet/worklet_thread_test_common.cc
index 5ece22c3ea7391b595fe441820a9a91a9b67ae1e..a6323f23329e6f72d544961f75932a14b00f2359 100644
--- a/third_party/blink/renderer/modules/worklet/worklet_thread_test_common.cc
+++ b/third_party/blink/renderer/modules/worklet/worklet_thread_test_common.cc
@@ -39,7 +39,7 @@ CreateAnimationAndPaintWorkletThread(
           window->GetReferrerPolicy(), DocumentPolicy::DocumentPolicyBundle{},
           window->GetSecurityOrigin(), window->IsSecureContext(),
           window->GetHttpsState(), clients,
-          nullptr /* content_settings_client */,
+          nullptr /* content_settings_client */, window->AddressSpace(),
           OriginTrialContext::GetInheritedTrialFeatures(window).get(),
           base::UnguessableToken::Create(), nullptr /* worker_settings */,
           mojom::blink::V8CacheOptions::kDefault,
